spark3/src/main/java/org/apache/iceberg/spark/FileRewriteCoordinator.java

```diff
@@ -0,0 +1,169 @@
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableSet;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.Pair;
+import org.apache.iceberg.util.Tasks;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FileRewriteCoordinator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(FileRewriteCoordinator.class);
+  private static final FileRewriteCoordinator INSTANCE = new FileRewriteCoordinator();
+
+  private final Map<String, Lock> lockMap = Maps.newConcurrentMap();
+  private final Map<Pair<String, String>, Set<DataFile>> resultMap = Maps.newConcurrentMap();
+
+  private FileRewriteCoordinator() {
+  }
+
+  public static FileRewriteCoordinator get() {
+    return INSTANCE;
+  }
+
+  public void stageRewrite(Table table, String fileSetID, Set<DataFile> newDataFiles) {
+    Preconditions.checkArgument(newDataFiles != null && newDataFiles.size() > 0, "Cannot stage null or empty files");
+    LOG.info("Staging results of rewriting file set {} for table {}", fileSetID, table);
+    Pair<String, String> id = toID(table, fileSetID);
+    resultMap.put(id, newDataFiles);
+  }
+
+  public void commitRewrite(Table table, String fileSetID) {
+    commitRewrite(table, ImmutableSet.of(fileSetID));
+  }
+
+  public void commitRewrite(Table table, Set<String> fileSetIDs) {
+    LOG.info("Committing rewrite of file sets {} for table {}", fileSetIDs, table);
+
+    Set<DataFile> rewrittenDataFiles = fetchRewrittenDataFiles(table, fileSetIDs);
+    Set<DataFile> newDataFiles = fetchNewDataFiles(table, fileSetIDs);
+
+    String tableUUID = tableUUID(table);
+    Lock tableLock = lockMap.computeIfAbsent(tableUUID, uuid -> new ReentrantLock());
+
+    try {
+      LOG.info("Locking table {} to commit rewrite of file sets {}", table, fileSetIDs);
+      tableLock.lock();
+
+      table.newRewrite()
+          .rewriteFiles(rewrittenDataFiles, newDataFiles)
+          .commit();
+
+      LOG.info("Successfully committed rewrite of file sets {} for table {}", fileSetIDs, table);
+
+    } finally {
+      LOG.info("Unlocking table {}", table);
+      tableLock.unlock();
+    }
+  }
+
+  private Set<DataFile> fetchRewrittenDataFiles(Table table, Set<String> fileSetIDs) {
+    FileScanTaskSetManager taskSetManager = FileScanTaskSetManager.get();
+
+    Set<DataFile> rewrittenDataFiles = Sets.newHashSet();
+
+    for (String fileSetID : fileSetIDs) {
+      List<FileScanTask> tasks = taskSetManager.fetchTasks(table, fileSetID);
+      ValidationException.check(tasks != null,
+          "Task set manager has no tasks for table %s with id %s",
```

Comment:
I think this and the validation below should probably just share the same message

Reply:
Copied it from `SparkFilesScan` for consistency.